File: card.py
class Card:
    """
    Represents a playing card.

    Attributes:
        rank (str): The rank of the card.
        suit (str): The suit of the card.
    """
    ranks = [_ for _ in range(2,11)] + ['JACK', 'QUEEN', 'KING', 'ACE']
    suits = ['SPADE', 'HEART', 'DIAMOND', 'CLUB']

    def __init__(self, rank, suit):
        self.rank = rank
        self.suit = suit

    def __str__(self):
        """
        Returns a string representation of the card.

        The string representation includes the rank and suit of the card.
        If the rank is a special case (J, Q, K, A), it is replaced with the corresponding word (Jack, Queen, King, Ace).

        Returns:
            str: The string representation of the card.
        """
        case = {
            'J': 'Jack',
            'Q': 'Queen',
            'K': 'King',
            'A': 'Ace'
        }
        rank = self.rank
        rank = case[rank] if rank in case else rank
        return f'{rank} of {self.suit}S'

    @property
    def rank(self):
        """
        Returns the rank of the card.

        Returns:
            str: The rank of the card.
        """
        return self._rank

    @rank.setter
    def rank(self, rank):
        """
        Set the rank of the card.

        Args:
            rank (str): The rank of the card.

        Raises:
            ValueError: If the rank is not a valid rank.

        Returns:
            None
        """
        if rank in Card.ranks:
            self._rank = rank
        else:
            raise ValueError('Invalid rank')

    @property
    def value(self):
        """
        Returns the value of the card.

        Returns:
            int: The value of the card.
        """
        if self.rank in Card.ranks[0:-4]:
            return int(self.rank)
        elif self.rank == 'ACE':
            return 11
        else:
            return 10

    @property
    def suit(self):
        """
        Returns the suit of the card.

        Returns:
            str: The suit of the card.
        """
        return self._suit

    @suit.setter
    def suit(self, suit):
        """
        Set the suit of the card.

        Parameters:
        - suit (str): The suit of the card. Must be one of ['Hearts', 'Diamonds', 'Clubs', 'Spades'].

        Raises:
        - ValueError: If the suit is not a valid option.

        """
        if suit in Card.suits:
            self._suit = suit
        else:
            raise ValueError('Invalid suit')

File: src/js/card.js
/**
 * Array of card ranks.
 * @type {string[]}
 */
export const RANKS = ['A', '2', '3', '4', '5', '6', '7', '8', '9', '10', 'J', 'Q', 'K'];
/**
 * Array of card suits.
 * @type {string[]}
 */
export const SUITS = ['SPADES', 'HEARTS', 'CLUBS', 'DIAMONDS'];

/**
 * Represents a playing card.
 * @class
 */
class Card {
  /**
   * The width of the card frame.
   * @type {number}
   */
  static FRAME_WIDTH = 88;
  /**
   * The height of the card frame.
   * @type {number}
   */
  static FRAME_HEIGHT = 124;

  /**
   * Represents a playing card.
   * @class
   * @param {string} rank - The rank of the card.
   * @param {string} suit - The suit of the card.
   */
  constructor(rank, suit) {
    this._rank = rank;
    this._suit = suit;
    this._faceUp = true;
  }

  /**
   * Get the rank of the card.
   *
   * @returns {string} The rank of the card.
   */
  get rank() {
    return this._rank;
  }

  /**
   * Set the rank of the card.
   *
   * @param {string} rank - The rank of the card.
   * @throws {Error} If the provided rank is invalid.
   */
  set rank(rank) {
    if (!RANKS.includes(rank)) {
      throw new Error(`Invalid rank: ${rank}`);
    }

    this._rank = rank;
  }

  /**
   * Get the suit of the card.
   *
   * @returns {string} The suit of the card.
   */
  get suit() {
    return this._suit;
  }

  /**
   * Set the suit of the card.
   *
   * @param {string} suit - The suit of the card.
   * @throws {Error} If the provided suit is invalid.
   */
  set suit(suit) {
    if (!SUITS.includes(suit)) {
      throw new Error(`Invalid suit: ${suit}`);
    }

    this._suit = suit;
  }

  flip() {
    this.faceUp = !this.faceUp;
  }

  /**
   * Checks if the card is face up.
   * @returns {boolean} True if the card is face up, false otherwise.
   */
  isFaceUp() {
    return this.faceUp;
  }

  /**
   * Get the current face-up status of the card.
   *
   * @returns {boolean} The face-up status of the card.
   */
  get faceUp() {
    return this._faceUp;
  }

  /**
   * Setter for the faceUp property.
   *
   * @param {boolean} faceUp - The new value for the faceUp property.
   */
  set faceUp(faceUp) {
    this._faceUp = faceUp;
  }

  /**
   * Get the value of the card.
   * @returns {number} The value of the card.
   */
  get value() {
    if (this.rank in RANKS.slice(0, -4)) {
      return parseInt(this.rank);
    }

    if(this.rank === 'A') {
      return 11;
    }

    return 10;
  }

  /**
   * Returns a string representation of the card.
   * @returns {string} The string representation of the card.
   */
  toString() {
    let labels = {
        'J': 'Jack',
        'Q': 'Queen',
        'K': 'King',
        'A': 'Ace'
    };

    return `${this.rank in labels ? labels[this.rank] : this.rank} of ${this.suit}s`;
  }

  /**
   * Returns the sprite offset for the card.
   * @returns {Object} The sprite offset object with x and y coordinates.
   */
  getSpriteOffset() {
    const x = RANKS.indexOf(this.rank);
    const y = SUITS.indexOf(this.suit);
    return {x, y};
  }
}

export default Card;